### PR TITLE
Add Enum field

### PR DESCRIPTION
**Pull request by lafrech**
Replaces @orenc17's #1885.

This is a long standing feature request. See for instance discussions in #267.

The status quo is that people can use @justanr's [marshmallow_enum](https://github.com/justanr/marshmallow_enum).

Currently, this lib doesn't seem actively maintained. I don't blame the author. A few things are outdated, deprecated stuff. The changes since MA3 have been user contributed and some are still missing.

I'm thinking if we recommend this lib as the reference choice, why not integrate it in code? This is what I've been trying to do.

I started by importing the code and updating it for PY3/MA3.

I also removed the dump/load asymmetry (dump by value, load by name) that I found awkward and revamped the error message generation.

I was still not happy with the `by_value`/`by_name` mechanism. I serialize by name, so I could live with only this, but I understand the need for serializing by value (to get an int or a nicer string). However, I like the type to be well-defined so that we can inherit deserialization from basic fields, and to help documenting the type (e.g. in apispec).

My proposal is therefore an `Enum` field serializing by name, and typed enum fields `StringEnum` and `IntegerEnum`, to serialize by value with a given type. Users would be free to create new types but I believe those cover most cases.

Transition from `marshmallow_enum` should be relatively smooth for users who don't use the dump/load asymmetry and who don't rely on exact error messages.

This still lacks a bit of doc but I'd rather get feedback before polishing.

Note we can't use `OneOf` validator because validation occurs after deserialization and wrong values are caught at deserialization already.

### CODE REVIEW

File: src/marshmallow/fields.py
Comment (deckar01) [inline src/marshmallow/fields.py:1922]: Tangent: This pattern is repeated in nearly every field. Is there any reason this isn't in `Field.serialize`? The complimentary `None` handling is implemented in `Field.deserialize`. Maybe something to tack onto #2012.

File: src/marshmallow/fields.py
Comment (deckar01) [inline src/marshmallow/fields.py:1928]: The main difference between `Enum` and `TypedEnum` seems to be that the former expects enum symbols, while the latter expects enum values. The field names don't currently communicate that.

Comment (lafrech) [reply src/marshmallow/fields.py:1928]: Well `TypedEnum` is not meant to be used, I just didn't make it private to allow for other user-defined typed enums. So it is more about `Enum` vs. `StringEnum` and `IntegerEnum`.

But I agree the names don't really convey the meaning. I'm open to suggestions. Docstrings will help clarifying.

Comment (deckar01) [reply src/marshmallow/fields.py:1928]: I would probably avoid preferring one use case over another with the shorthand `Enum`. `EnumSymbol` and `EnumValue` would be clear regardless of what the user's intuition is.

Comment (lafrech) [reply src/marshmallow/fields.py:1928]: Why not?

So should it be

- `EnumSymbol`
- `EnumValue`
- `EnumString`
- `EnumInteger`

?

Or

- `EnumSymbol`
- `EnumValue`
- `StringEnumValue`
- `IntegerEnumValue`

?

I don't like the former. Kinda like the latter.

Comment (justanr) [reply src/marshmallow/fields.py:1928]: Latter is clearer to me.

File: tests/base.py
Comment (justanr) [inline tests/base.py:20]: It's me (:

File: src/marshmallow/fields.py
Comment (justanr) [inline src/marshmallow/fields.py]: If I was doing everything all over again, I'd leave it at just the string representation of the enum and call it a day. The only value with using an underlying numeric is that when sending to some other languages they can successfully decode an unknown value into their enum type. But that just seems like asking for subtle bugs instead of being hit with "Cannot deserialize MyEnum.FooBar15" or whatever they'd spit out. I've been away from python long enough that I don't recall what Python does off hand but if I had to guess Python barfs when attempting that.

**Comment by justanr** [comment]
Funnily enough, I showed up here because I got notified I'm a maintainer of a critical package, which turns out is the one I  question.

File: src/marshmallow/fields.py
Comment (lafrech) [reply src/marshmallow/fields.py]: Rather than `str(value)`, which would expose the enum name, I'd rather use the member name (`SymbolEnum`). Offering to serialize by value allows users to provide human-readable names as enum values, for instance, or at least decouple business logic and API layer a bit.

But thanks for the feedback. Your comment confirms my intuition that the dump/load asymmetry is too much API surface for the need.

File: tests/base.py
Comment (lafrech) [reply tests/base.py:20]: Yeah, I figured while touching this part of the code I might as well make it more inclusive.

I hope no one will complain about the hair colour choice being too restrictive.

File: src/marshmallow/fields.py
Comment (justanr) [reply src/marshmallow/fields.py]: Yeah the symbol name is what I meant, just worded it kind of poorly. The asymmetry was a pain in the ass for tests and it tripped me up a few times. Pretending that they're just a union of string literals is the way to go imo

**Comment by lafrech** [comment]
Renamed  Enum -> EnumSymbol, TypedEnum -> EnumValue. I'll squash later on.

Docstrings still missing.

File: src/marshmallow/fields.py
Comment (deckar01) [reply src/marshmallow/fields.py:1928]: What about `EnumValue(String)` and `EnumValue(Integer)`? That is consistent with `Nested`/`List` API, improves readability, and minimizes the API surface area.

Comment (lafrech) [reply src/marshmallow/fields.py:1928]: Good idea. I shall investigate this.

File: tests/base.py
Comment (deckar01) [reply tests/base.py:20]: Hair color should be a `Flag` so that you can dye multiple colors. 😛

**Comment by lafrech** [comment]
> What about `EnumValue(String)` and `EnumValue(Integer)`? That is consistent with `Nested`/`List` API, improves readability, and minimizes the API surface area.

Just tried this and it worked a treat!

I'm happy with this implementation.

Still requires docstrings and perhaps testing with a DateTime field or some other field for which ser/deser is less trivial to assert ser/deser actually happens.

**Comment by lafrech** [comment]
I think I'm done.

@sloria @deckar01 @justanr I'll wait for a few days in case someone wants to review this and then I'll merge.

**Review by deckar01 — APPROVED** [review]

**Review by justanr — APPROVED** [review]

File: src/marshmallow/fields.py
Comment (lafrech) [inline src/marshmallow/fields.py]: This is wrong, we need to use the field to dump the value. `str` only work for trivial cases.

**Review by lafrech — COMMENTED** [review]
As PR author, I can't request changes.

Please don't merge until I fix.

File: src/marshmallow/fields.py
Comment (lafrech) [reply src/marshmallow/fields.py]: I just pushed a change to let the field serialize the values.

Our API kinda sucks because serialize expects `attr` and `obj` parameters and we don't have those and actually I don't think they are ever used. There might be a nice opportunity for a refactor here. I opened #2039.

**Comment by lafrech** [comment]
PR fixed. I'm not 100% happy with `self.field._serialize(m.value, None, None)` but that's our API.

File: src/marshmallow/fields.py
Comment (lafrech) [reply src/marshmallow/fields.py]:
```py
self.field._serialize(m.value, None, None)
```

This won't work for fields that need `attr` or `obj` (but is there any ? #2039) or fields that are not symmetric (don't dump what they loaded). I'm not talking about fields that don't round-trip, I mean if several equivalent representation are accepted at load time and only one is used at dump time, this one is the one that will be displayed here and that's fine, I believe. It would be more of a problem for custom field that would purposely have asymmetric load/dump methods, like using different time formats, for instance.

Since those must be quite rare, and using them in this enum field would be even more rare, I'm fine with this limitation.

We could remove the list of accepted values from the error message but I think it is useful in the "normal" case (99% of the time) and it will be needed for documentation purpose anyway (apispec).

Comment (lafrech) [reply src/marshmallow/fields.py]:
```py
self.field._serialize(m.value, None, None)
```

 is already used in `Mapping`.

**Comment by lafrech** [comment]
Alright, moving on with this.

**Comment by lafrech** [comment]
Just after merging this, as I was updating the changelog to release, I realized the naming didn't follow our habits of naming fields by object type.

I pushed a new PR to propose merge of both fields into an `Enum` field: #2044.

**Comment by ThiefMaster** [comment]
Would you consider adding a short migration guide for `marshmallow_enum` users? :)

**Pull request merged**
